feat: v3 api support

danya opened the pull request.

kulti committed:
feat: v3 api support

danya committed:
fix: use v3 as API prefix

danya committed:
fix: invalid path for devices commands

kulti committed:
chore: rename v3 version binary to enapter3

kulti committed:
chore: change command keys format (from _ to -)

kulti committed:
feat: add profiles download/upload commands

kulti committed:
refactor,tests: don't compare helps in update case

kulti committed:
tests: add http req tests

kulti committed:
feat: name for blueprint archive as *.enbp

kulti committed:
feat: support rule-engine commands

kulti committed:
feat: require name for rule, not slug

kulti committed:
feat: set req body content type

kulti committed:
feat: add verbose mode

kulti committed:
bug: clean path before zip dir

kulti committed:
refactor: move device-id option into devices subcommand

kulti committed:
feat: fix devices command names

kulti committed:
feat: fix provisioning command names

kulti committed:
feat: fix blueprints command names

kulti committed:
feat: add device inspect and list commands

kulti committed:
feat: remove blueprint commands position args

rnovatorov committed:
fix: remove redundant print

anataty committed:
feat: remove name field from rule

rnovatorov committed:
rule-engine: fix create and update commands

danya committed:
feat: support streaming of rule logs

rnovatorov commented on internal/app/enaptercli/cmd_rule_engine_rule_logsf.go:
```diff
@@ -0,0 +1,70 @@
+	return append(c.cmdRuleEngineRule.Flags(),
+		&cli.StringFlag{
+			Name:        "rule-id",
+			Usage:       "Rule ID or slug to update",
```
Usage description is invalid. Should be something like `ID or slug of a rule to stream logs of`.

rnovatorov commented on internal/app/enaptercli/testdata/helps/enapter rule-engine rule:
```diff
@@ -13,6 +13,7 @@ COMMANDS:
    list           List rules
    update         Update a rule
    update-script  Update the script of a rule
+   logsf          Stream rule logs
```
It may be wise to name the command `logs` and add a required flag `-f, --follow` to be able to extend the command with more flags later on and not force `f` semantics. We might for example want to add time filters. Take a look at how Docker CLI does this:

```shell
$ docker logs --help

Usage:  docker logs [OPTIONS] CONTAINER

Fetch the logs of a container

Aliases:
  docker container logs, docker logs

Options:
      --details        Show extra details provided to logs
  -f, --follow         Follow log output
      --since string   Show logs since timestamp (e.g. "2013-01-02T13:23:37Z") or relative (e.g. "42m" for 42 minutes)
  -n, --tail string    Number of lines to show from the end of the logs (default "all")
  -t, --timestamps     Show timestamps
      --until string   Show logs before a timestamp (e.g. "2013-01-02T13:23:37Z") or relative (e.g. "42m" for 42 minutes)
```

danya replied:
We can switch to `logs --follow`, but there're a few moments.

* Rule logs would always require `--follow` flag.
* The output formats would differ: `logs` provides single JSON with pagination, while `logsf` returns multiline JSON wuithout pagination.

So, I'd prefer to keep `logsf` or rename it to `stream-logs` or something like that.

rnovatorov replied:
A required `-f` flag seems fine to me as we'll not break anything if we change it to be optional when we'll support other logs features.

As for pagination I'm not sure I follow what you mean here. How does pagination prevent us from having the same output format in both streaming and non-streaming version of `logs` command?

danya replied:
Oh, I see that we're post-processing the logs from the server, rather than just returning the API response like in other commands. Now, I think we can have both streaming and non-streaming versions of the same command.

kulti committed:
feat: provision lua device with a slug

danya committed:
feat: add reconnections for rule logs streaming

kulti committed:
bug: get nil req body in verbose mode

kulti committed:
feat: add device delete command

kulti committed:
feat: support device command executions

egakupi committed:
fix(rule-engine): validate runtime-version

danya committed:
fix: uppercase logs enums

danya committed:
feat: stream device logs via WebSocket

kulti committed:
feat: list devices from specified site

kulti committed:
feat: list and inspect sites

kulti committed:
feat: assign-blueprint and provision lua-device with blueprint from path

kulti committed:
tests: check request

kulti committed:
chore: bump go version to 1.23

kulti committed:
feat: rename inspect -> get

kulti committed:
feat: rename api-host to api-url and make it visible option

kulti committed:
feat: allow insecure http connection to api

kulti committed:
feat: add ENVIRONMENT VARIABLES to all helps

kulti committed:
feat: rework device command execution

kulti committed:
feat: cleanup unnecessary option prefixes

kulti committed:
fix: site-id is not required to provision standalone on Gateway

kulti committed:
feat: add missed short name for device-id option

kulti committed:
feat: hide ephemeral API

kulti committed:
docs: add expand supported values to help messages

kulti committed:
feat: generate device communication config

kulti committed:
feat: delete default exec-interval value

kulti committed:
fix: make site list close to API

kulti committed:
doc: update help messages

kulti committed:
feat: suggest --api-allow-insecure option

kulti committed:
feat: support non-json error response

kulti committed:
feat: support pagination for devices and sites lists

kulti committed:
feat: rename provisioning to 'device create'

kulti committed:
feat: rename assign-blueprint to change-blueprint

kulti committed:
bug: device generate communication config should be POST

kulti committed:
bug: create lua_device or change bluerint should handle blueprin upload errors

kulti committed:
doc: improve usage for duration flags

danya committed:
feat: support TLS for WebSockets

kulti committed:
feat: delete rule one by one

kulti committed:
chore: update ci

danya committed:
feat: write CLI logs to stderr

danya committed:
feat: improve WebSockets error handling

danya committed:
feat: add device telemetry command

kulti committed:
fix: update device expand values

kulti committed:
fix: rename new_rule_script to script when update rule

kulti committed:
help: add notice about device list order

danya committed:
feat: add device update command

danya committed:
refactor: change 'devices' to 'device'

danya committed:
Merge master into v3 branch

Resolve merge conflicts to enable CI checks.

Squashed commit of the following:

commit 36d65cfc4bf092fdbb9c2551e57620d73bd710f1
Author: Nikolay V. Krasko <nikolay@enapter.com>
Date:   Fri Sep 5 15:52:41 2025 +0400

    adding notes with status

commit 2277748fee99d94d4a374646a0b266377d4fba9c
Author: Nikolay V. Krasko <nikolay@enapter.com>
Date:   Fri Sep 5 15:43:37 2025 +0400

    addding enapter3 cli usage description

danya committed:
feat: print downloaded file name

rnovatorov committed:
cmd device change-blueprint: fix typo in description

danya committed:
feat: expect logs timestamp as int64

danya committed:
feat: add custom user agent header

danya committed:
chore: migrate to golangci-lint v2

rnovatorov committed:
fix tests of help messages

rnovatorov committed:
ci: run on `v3` branch

rnovatorov committed:
device: create-standalone: add `--device-slug` option

danya committed:
refactor: consistent use of the singular form

danya committed:
fix: broken paths on Windows

danya committed:
feat: add initial version of remote terminal

danya committed:
feat: add connection management

danya committed:
feat: use associated site-id from rule-engine commands

danya committed:
feat: use associated site-id for site commands

danya committed:
feat: use associated site-id for device commands

danya committed:
feat: use associated site-id for provisioning commands

danya committed:
test: add more tests

danya committed:
feat: new logs format

danya committed:
feat: improve missing site id error message

danya committed:
feat: capitalize flags descriptions

danya committed:
feat: trim device names

danya committed:
feat: add device monitor command

danya committed:
feat: add colors for logs

nkrasko committed:
adding reference for cli v3 and update readme

nkrasko committed:
fixing typo in readme

nkrasko committed:
i am infected with covid. sorry for another typo

danya committed:
feat: improve TLS error description

danya committed:
feat: update logs API

danya committed:
feat: add ENAPTER3_CONNECTION

kulti committed:
refactor: migrate to urfave/cli/v3

kulti committed:
feat: support .blueprintignore

kulti committed:
feat: public go package blueprint

kulti committed:
feat: trim spaces in names

kulti committed:
chore: go module should be v3

rnovatorov committed:
feat: implement rule engine list pagination

Rule Engine API has introduced 'offset' and 'limit' parameters.
This update ensures 'enapter rule-engine rule list' supports
pagination by leveraging the common pagination engine.

danya committed:
fix: invalid path for rule logs streaming

rnovatorov committed:
fix: handle legacy rule engine pagination

rnovatorov committed:
refactor: use parseRespErrorMessage for detailed error reporting in pagination
